common/src/character_controller/vector_bounds.rs

```diff
@@ -0,0 +1,298 @@
+//! This module is used to transform vectors to ensure that they fit constraints discovered during collision checking.
+
+use rand_distr::num_traits::Zero;
+use tracing::warn;
+
+use crate::math;
+
+/// Encapsulates all the information needed to constrain a vector based on a set of `VectorBound`s.
+#[derive(Clone)]
+pub struct VectorBoundGroup {
+    bounds: Vec<VectorBound>,
+    error_margin: f32,
+}
+
+impl VectorBoundGroup {
+    /// Initializes a `VectorBoundGroup` with an empty list of bounds. The `initial_vector` is the first vector
+    /// we expect these bounds to be applied to, a hint to determine what kind of error margin is needed
+    /// to prevent floating point approximation limits from causing phantom collisions. Note that this
+    /// error margin is not needed if the resulting vector is zero, since no phantom collision can occur
+    /// if the character is stopped.
+    pub fn new(initial_vector: &na::Vector3<f32>) -> Self {
+        let error_margin = initial_vector.magnitude() * 1e-4;
+
+        VectorBoundGroup {
+            bounds: vec![],
+            error_margin,
+        }
+    }
+
+    /// Constrains `vector` with `new_bound` while keeping the existing constraints satisfied.
+    /// All projection transformations applied to `vector` are also applied
+    /// to `tagalong` to allow two vectors to be transformed consistently with each other.
+    pub fn apply_and_add_bound(
+        &mut self,
+        new_bound: VectorBound,
+        vector: &mut na::Vector3<f32>,
+        tagalong: Option<&mut na::Vector3<f32>>,
```

Comment:
Possibly. The issue is that we have to extract the values out at the end if they're fields, and the names `vector` and `tagalong` aren't self-explanatory, causing a potential readability pitfall, and, we'd need an `unwrap` when extracting `tagalong` at the end.

Both of these could potentially be addressed with traits, but that might hurt readability even more.

Reply:
A trait isn't justified here.

The names aren't any more confusing as fields than as arguments, IMO. Still, they could be improved regardless. Could we name them for what they're actually used for in practice? It's not like we have wildly diverse callsites.

Reply:
Currently, there are two usages: `apply_velocity` uses it to update a velocity, and `get_ground_normal` uses it to update a displacement. I believe using a specific field name might prove difficult.

However, switching around the order of some multiplications etc. should be able to allow these two usages to be made more similar to each other, which is not something I had considered when I previously attempted to make them variables.

Another difficulty I had in my last attempt is that I had other methods that manipulated these vectors, causing borrow-check issues under certain situations. A lot has changed since then, so I'm reasonably optimistic about your suggestion.

I'm always happy to reduce the parameter count.